Source/SwiftLintFramework/Models/RuleStorage.swift

```diff
@@ -0,0 +1,25 @@
+import Dispatch
+import SourceKittenFramework
+
+public class RuleStorage {
+    private var storage: [ObjectIdentifier: [File: Any]]
+    private let access = DispatchQueue(label: "io.realm.swiftlint.ruleStorageAccess", attributes: .concurrent)
+
+    public init() {
+        storage = [:]
+    }
+
+    func collect<R: CollectingRule>(info: R.FileInfo, for file: File, in rule: R) {
+        let key = ObjectIdentifier(R.self)
+        access.sync(flags: .barrier) {
+            storage[key, default: [:]][file] = info
+        }
+    }
+
+    func collectedInfo<R: CollectingRule>(for rule: R) -> [File: R.FileInfo] {
+        return access.sync {
+            // swiftlint:disable:next force_cast
+            storage[ObjectIdentifier(R.self)] as! [File: R.FileInfo]
```

Comment:
Can you explain to me why this is a safe use of force casting? If we fail to run `collect` with a rule, then call `collectedInfo` for that rule, this will crash, without printing a nice error message to the user. They'll just see a segfault.

Why not default to an empty dictionary?

```suggestion
            storage[ObjectIdentifier(R.self)] as? [File: R.FileInfo] ?? [:]
```

Reply:
IIRC, I wanted to crash if the collected info for a file was being read from before it was written to. This should never happen and indicates that something's gone wrong internally.

But looking at it now, the force-cast doesn't make that assertion clear. If you agree that crashing is here okay, I can make this return value optional and put a precondition in `CollectingRule.validate(file:using:compilerArguments:)`. If you want to avoid a crash, defaulting to an empty dictionary makes sense to me 🙂

Reply:
A precondition or assertion failure makes sense to me. I'd just rather have informative crashes 😄

Reply:
Sounds good! Done in https://github.com/realm/SwiftLint/pull/2714/commits/7bd359e69f8b969921123590c0cd1cb9e84ba567